pkgs/development/haskell-modules/generic-builder.nix

```diff
@@ -10,12 +10,81 @@
   runCommandCC,
   ghcWithHoogle,
   ghcWithPackages,
+  haskellLib,
+  iserv-proxy,
   nodejs,
+  writeShellScriptBin,
 }:
 
 let
   isCross = stdenv.buildPlatform != stdenv.hostPlatform;
 
+  crossSupport = rec {
+    emulator = stdenv.hostPlatform.emulator buildPackages;
+
+    # This is a silly workaround to avoid infinite recursion.
+    # We don't know which packages need template haskell, so we always need iserv-proxy for the host
+    # but then that package also has dependencies which would themselves need iserv-proxy.
+    # TODO: grab host iserv-proxy from the "original" package set without TH support, then have host TH support on the "real" package set.
+    iservProxyDependency =
+      pname:
+      builtins.elem pname [
+        "iserv-proxy"
+        "libiserv"
+        "network"
+      ];
+
+    hasBuiltinTH = stdenv.hostPlatform.isGhcjs;
+
+    canProxyTH =
+      pname:
+      lib.versionAtLeast ghc.version "9.6"
+      && stdenv.hostPlatform.emulatorAvailable buildPackages
+      && !(crossSupport.iservProxyDependency pname);
+
+    # Many suites use Template Haskell for test discovery, including QuickCheck
+    canCheck = pname: hasBuiltinTH || canProxyTH pname;
+
+    iservWrapper =
+      let
+        buildProxy = iserv-proxy.build + "/bin/iserv-proxy";
+
+        wrapperScript =
+          enableProfiling:
+          let
+            overrides = haskellLib.overrideCabal {
+              enableLibraryProfiling = enableProfiling;
+              enableExecutableProfiling = enableProfiling;
+            };
+            hostProxy = overrides iserv-proxy.host + "/bin/iserv-proxy-interpreter";
+          in
+          buildPackages.writeShellScriptBin ("iserv-wrapper" + lib.optionalString enableProfiling "-prof") ''
+            set -euo pipefail
+            PORT=$((5000 + $RANDOM % 5000))
+            (>&2 echo "---> Starting interpreter on port $PORT")
+            ${emulator} ${hostProxy} tmp $PORT &
+            (>&2 echo "---| interpreter should have started on $PORT")
```

Comment:
This log message is pretty useless, I think, because it will never wait for anything that the preceding line does. The `&` detaches, so the echo will happen immediately.. even if the interpreter failed to start.